### ISSUE TITLE
Broken typechecking with generics type at classes

### ISSUE THREAD

**Issue by lordent** [microsoft/TypeScript#52938]
# Bug Report

- Type checking is skipped
- v4.9.5

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/EQVwzgpgBGAuBOBLAxrYAodEAeAHA9vLFLAJ67QAqAsqQKIA2EAthAHawDyusi+bYKAF4oAb3RRJUAIawEYAPwAuMRKnrkDaWDAA5aa2UwEiNgHM1kgL7obWPIWJkKUGvSasOAEVnThYu01tQVpGFnZYAB5LKG5efjBKcmgcWHYAE0E3MM8uHj4Bf2yPCLiCsHQAPihUjMEACRoAGRyI1RjkBIQQVEIACnx8hKMyhKSKAEpVdXUwEAp4PonMGalYAAtEMAA6U0RYAaGBBW3ZeWX1O3U9g7P4RRUAJQgARxBEeAh0yNGBcYhKgBtADkdzAwIAulNxDEbHYcAQiCRkq5aABJTpsHywPwicQafDpCBGOBIcwdfAMJioAokkzk+EOJHOKi0ABCIDk-GxuNR7nC3l8UAAZNMpCh+EY3Bjub4KRwInSyRZGYinCi3ByuWxfoIRMUBXl4oVRfipGCpezObBZTjbJggjooFbtdF1Lr-jVsGk2Jk+VqbTqjnr-db+Lqql6fX7QiUOD9g-9qqIoIEulAAEZhtj+NgQADuztIAf4fTNkjBKnLMwAZvh8CoAMwAFgATC2ADQxWb4VgqboQLurSSYn2wFTAtEyZiZ7PAodSOHLdCYuAyOT3VtPV7vT7fTXZ3VA0Eb8EQ-zVusNqAt9vNheSMC9iD9+AgQfyscTqcGWfa+f2kAA)

### 💻 Code

```ts
"use strict"

export type TMyElementOptions = {
    attrs?: {
        className?: string
    }
}

export type TMyElementData = {}

class MyElement<
    OptionsType extends TMyElementOptions = TMyElementOptions
> extends HTMLElement {

    constructor(options?: OptionsType) {
        super()

        this.init(options?.attrs)
    }

    init(attrs?: Required<OptionsType>['attrs']) {

    }
}

export type TMyIconData = {
    code?: string
    collection?: string
}

export type TMyButtonData = TMyElementData & {
    icon?: TMyIconData
    content?: string
}

export type TMyButtonOptions = TMyElementOptions & {
    attrs?: TMyButtonData
}

class MyButton<
    OptionsType extends TMyButtonOptions = TMyButtonOptions
> extends MyElement<OptionsType> { }

const button = new MyButton({
    attrs: {
        foo: 34234, /* HERE */
        some: true, /* HERE */
        content: 'I am button',
    }
})

const attrs2: Required<TMyButtonOptions>['attrs'] = {
    foo: 34234,  /* WORKING! */
    some: true,
    content: 'I am button',
}

```

### 🙁 Actual behavior

Logically, type checking should return an error when incorrect parameters are specified in the constructor, but this does not happen.

### 🙂 Expected behavior

I expect that the type checking of the arguments passed to the constructor will be symmetrical to the type checking specified directly, as in the attrs2 example

**Comment by MartinJohns** [comment]
Why do you believe this should be a type error? The type argument for `MyButton` is inferred based on the argument provided. The argument you pass matches the type constraint, so it's allowed.

**Comment by smiren** [comment]
@MartinJohns Don't understand your passage. Why do you belive that passed argument matches type constraint?

**Comment by MartinJohns** [comment]
- The type argument of `MyButton<>` is defined as `OptionsType extends TMyElementOptions`, so whatever type is used, it must extend `TMyElementOptions`.
- `TMyElementOptions` is typed as `{ attrs?: { className?: string } }`.
- The type inferred in line 44 based on the argument is `{ attrs: { foo: number; some: boolean; content: string } }`.
- The type `{ attrs: { foo: number; some: boolean; content: string } }` does extend `TMyElementOptions`, because everything that must be there according to `TMyElementOptions` is present in the inferred type, and there's nothing making these two types incompatible.

**Comment by fatcerberus** [comment]
It's maybe a bit curious that this isn't caught by [weak type checks](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-4.html#weak-type-detection) though, fwiw

**Comment by RyanCavanaugh** [comment]
MartinJohns is correct. You wrote
```ts
class MyButton<
    OptionsType extends TMyButtonOptions = TMyButtonOptions
> extends MyElement<OptionsType> { }

const button = new MyButton({
    attrs: {
        foo: 34234, /* HERE */
        some: true, /* HERE */
        content: 'I am button',
    }
})
```
instead of
```ts
class MyButton extends MyElement<TMyButtonOptions > { }

const button = new MyButton({
    attrs: {
        foo: 34234, /* HERE */
        some: true, /* HERE */
        content: 'I am button',
    }
})
```

Allowing a subtype of the specified type is what it *means* to be generic; it doesn't make any sense to error here.

**Comment by fatcerberus** [comment]
@RyanCavanaugh weak type check should make it *not* a subtype though?

**Comment by RyanCavanaugh** [comment]
Rejection due to weak type checking requires zero overlap in properties between a source type and all-optional target type. In this case, `content` overlaps. If you remove that property, then an error appears like you're expecting.

**Comment by fatcerberus** [comment]
Ah, I see - I was looking at `TMyElementOptions` which only defines `className`.  Somehow I missed that `MyButton` uses a more specific type.

**Comment by lordent** [comment]
> Rejection due to weak type checking requires zero overlap in properties between a source type and all-optional target type. In this case, `content` overlaps. If you remove that property, then an error appears like you're expecting.

Thanks for the answer [Ryan](https://github.com/RyanCavanaugh), you managed to explain how it works, but the problem is that it does not say how to solve problems in such my case.

My goal is to write a classes that will be reused in the future and should be open for extension
I can't write something like that, because the typescript doesn't know how to work like that, moreover, in the hint to the solution, it writes using extends syntax, so what should to do?

```ts
class MyElement<OptionsType = TMyElementOptions> extends HTMLElement {

    constructor(options?: OptionsType) {
        super()

        this.init(options?.attrs)  /* Property 'attrs' does not exist on type 'NonNullable<OptionsType>'.(2339) */
    }

    init(attrs?: OptionsType['attrs']) { /* Type '"attrs"' cannot be used to index type 'OptionsType'.(2536) */

    }
}

/* or this */

class MyButton<OptionsType = TMyButtonOptions> extends MyElement<OptionsType> { } /* Type 'OptionsType' does not satisfy the constraint 'TMyElementOptions'.(2344)
input.tsx(40, 16): This type parameter might need an `extends TMyElementOptions` constraint. */

```

**Comment by RyanCavanaugh** [comment]
I don't think you understand what generic defaults do. Anyway, this isn't a support forum.
